nix/eval-machine-info.nix

```diff
@@ -11,104 +11,170 @@
+                          lib.mapAttrs
+                            (nodeName: node:
+                              lib.mapAttrs
+                                (key: lib.warn "Resource ${r.name} accesses nodes.${nodeName}.${key}, which is deprecated. Use the equivalent option instead: nodes.${nodeName}.${newOpt key}.")
+                                config.nodes.${nodeName})
+                            config.nodes;
+                      };
+                      imports = [
+                        mainModule
+                        deploymentInfoModule
+                        ./resource.nix
+                      ];
+                    }));
+                  };
+                in {
+                  sshKeyPairs = resOpt ./ssh-keypair.nix;
+                  commandOutput = resOpt ./command-output.nix;
+                };
+                config = {
+                  machines = config.nodes;
+                  _module.check = false;
+                };
+              }] ++ pluginResources;
+              specialArgs = {
+                inherit evalResources resourcesByType;
+                inherit (lib) zipAttrs;
+              };
+            };
+          };
+          # Compute the definitions of the machines.
+          nodes = mkOption {
+            description = "The NixOS configurations for the nodes in the network.";
+            type = types.attrsOf (import "${nixpkgs}/nixos/lib/eval-config.nix" {
+              inherit lib system;
+              specialArgs = {
+                inherit uuid deploymentName;
+                inherit (config) nodes resources;
+              } // config.network.nodesExtraArgs;
+              modules = # Make NixOps's deployment.* options available.
+                pluginOptions ++
+                [
+                  ./options.nix
+                  ./resource.nix
+                  deploymentInfoModule
+                  config.defaults
+                  ({ name, ... }: rec{
+                    _file = ./eval-machine-info.nix;
+                    key = _file;
+                    # Provide a default hostname and deployment target equal
+                    # to the attribute name of the machine in the model.
+                    networking.hostName = lib.mkOverride 900 name;
+                    deployment.targetHost = lib.mkOverride 900 name;
+                    environment.checkConfigurationOptions = lib.mkOverride 900 checkConfigurationOptions;
+                    nixpkgs.system = lib.mkDefault system;
+                  })
+                ];
+            }).type;
+          };
+          defaults = mkOption {
+            type = types.anything;
+            default = { };
+            description = ''
+              Extra NixOS options to add to all nodes.
+            '';
+          };
+        };
+        config = let
+          nodes = removeAttrs config (builtins.attrNames options);
+        in lib.mkIf ({} != nodes) { #TODO: actual warning/assert module impl.
```

Comment:
Module system built-in functionality would be best (https://github.com/NixOS/nixpkgs/pull/97023#issuecomment-1059445608) but hasn't been implemented yet. It's a little tricky with strictness, but still feasible.